feat(api): add endpoint for fetching Game Leaderboards

ioslife opened the pull request:
```
/*
 *  API_GetGameLeaderboards - returns a list of Leaderboards for the give GameID
 *    i : gameID
 *    o : offset - number of entries to skip (default: 0)
 *    c : count - number of entries to return (default: 100, max: 500)
 *  int         Count                       number of want to play game records returned in the response
 *  int         Total                       number of want to play game records the user actually has overall
 *  array       Results
 *   object      [value]
 *    int        ID                         unique identifier of the leaderboard
 *    string     Title                      name of the leaderboard
 *    int        Description                details about what the leaderboard is tracking
 */
```

ioslife commented:
api-docs PR: https://github.com/RetroAchievements/api-docs/pull/54

ioslife commented on public/API/API_GetGameLeaderboards.php:
```diff
@@ -0,0 +1,56 @@
+<?php
+
+/*
+ *  API_GetGameLeaderboards - returns a list of Leaderboards for the give GameID
```
```suggestion
 *  API_GetGameLeaderboards - returns a list of Leaderboards for the given GameID
```

ioslife commented on public/API/API_GetGameLeaderboards.php:
```diff
@@ -0,0 +1,56 @@
+ *    i : gameID
+ *    o : offset - number of entries to skip (default: 0)
+ *    c : count - number of entries to return (default: 100, max: 500)
+ *  int         Count                       number of want to play game records returned in the response
+ *  int         Total                       number of want to play game records the user actually has overall
```
```suggestion
 *  int         Count                       number of leaderboard records returned in the response
 *  int         Total                       number of leaderboard records the game actually has overall
```

ioslife commented:
I'd like to add the current leaderboard winner to this before we merge it in

wescopeland commented:
> I'd like to add the current leaderboard winner to this before we merge it in

Let's move this to draft until it's ready for review.

ioslife commented:
> > I'd like to add the current leaderboard winner to this before we merge it in
> 
> Let's move this to draft until it's ready for review.

Yup, meant to do that when I had this after thought. TY for changing.

wescopeland commented:
When I call this endpoint, I get:
```json
{
	"message": "Attempt to read property \"user\" on null",
	"exception": "ErrorException",
	"file": "/var/www/html/public/API/API_GetGameLeaderboards.php",
	"line": 65,
	"trace": [
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php",
			"line": 255,
			"function": "handleError",
			"class": "Illuminate\\Foundation\\Bootstrap\\HandleExceptions",
			"type": "->"
		},
		{
			"file": "/var/www/html/public/API/API_GetGameLeaderboards.php",
			"line": 65,
			"function": "Illuminate\\Foundation\\Bootstrap\\{closure}",
			"class": "Illuminate\\Foundation\\Bootstrap\\HandleExceptions",
			"type": "->"
		},
		{
			"file": "/var/www/html/app/Http/Concerns/HandlesPublicFileRequests.php",
			"line": 46,
			"function": "require"
		},
		{
			"file": "/var/www/html/app/Api/RouteServiceProvider.php",
			"line": 90,
			"function": "handleRequest",
			"class": "App\\Api\\RouteServiceProvider",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/CallableDispatcher.php",
			"line": 40,
			"function": "App\\Api\\{closure}",
			"class": "App\\Api\\RouteServiceProvider",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php",
			"line": 237,
			"function": "dispatch",
			"class": "Illuminate\\Routing\\CallableDispatcher",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Route.php",
			"line": 208,
			"function": "runCallable",
			"class": "Illuminate\\Routing\\Route",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
			"line": 806,
			"function": "run",
			"class": "Illuminate\\Routing\\Route",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 144,
			"function": "Illuminate\\Routing\\{closure}",
			"class": "Illuminate\\Routing\\Router",
			"type": "->"
		},
		{
			"file": "/var/www/html/app/Api/Middleware/LogApiUsage.php",
			"line": 20,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "App\\Api\\Middleware\\LogApiUsage",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php",
			"line": 50,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Illuminate\\Routing\\Middleware\\SubstituteBindings",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Auth/Middleware/Authenticate.php",
			"line": 57,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Illuminate\\Auth\\Middleware\\Authenticate",
			"type": "->"
		},
		{
			"file": "/var/www/html/app/Http/Middleware/JsonResponse.php",
			"line": 16,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "App\\Http\\Middleware\\JsonResponse",
			"type": "->"
		},
		{
			"file": "/var/www/html/app/Api/Middleware/AccessControlAllowOriginWildcard.php",
			"line": 14,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "App\\Api\\Middleware\\AccessControlAllowOriginWildcard",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 119,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
			"line": 805,
			"function": "then",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
			"line": 784,
			"function": "runRouteWithinStack",
			"class": "Illuminate\\Routing\\Router",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
			"line": 748,
			"function": "runRoute",
			"class": "Illuminate\\Routing\\Router",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php",
			"line": 737,
			"function": "dispatchToRoute",
			"class": "Illuminate\\Routing\\Router",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php",
			"line": 200,
			"function": "dispatch",
			"class": "Illuminate\\Routing\\Router",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 144,
			"function": "Illuminate\\Foundation\\Http\\{closure}",
			"class": "Illuminate\\Foundation\\Http\\Kernel",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/livewire/livewire/src/Features/SupportDisablingBackButtonCache/DisableBackButtonCacheMiddleware.php",
			"line": 19,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Livewire\\Features\\SupportDisablingBackButtonCache\\DisableBackButtonCacheMiddleware",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/barryvdh/laravel-debugbar/src/Middleware/InjectDebugbar.php",
			"line": 66,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Barryvdh\\Debugbar\\Middleware\\InjectDebugbar",
			"type": "->"
		},
		{
			"file": "/var/www/html/app/Http/Middleware/FeatureFlagMiddleware.php",
			"line": 26,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "App\\Http\\Middleware\\FeatureFlagMiddleware",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/spatie/laravel-robots-middleware/src/RobotsMiddleware.php",
			"line": 14,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Spatie\\RobotsMiddleware\\RobotsMiddleware",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/spatie/laravel-missing-page-redirector/src/RedirectsMissingPages.php",
			"line": 12,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Spatie\\MissingPageRedirector\\RedirectsMissingPages",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php",
			"line": 21,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php",
			"line": 31,
			"function": "handle",
			"class": "Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Illuminate\\Foundation\\Http\\Middleware\\ConvertEmptyStringsToNull",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php",
			"line": 21,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php",
			"line": 40,
			"function": "handle",
			"class": "Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Illuminate\\Foundation\\Http\\Middleware\\TrimStrings",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php",
			"line": 27,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php",
			"line": 99,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Illuminate\\Foundation\\Http\\Middleware\\PreventRequestsDuringMaintenance",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/HandleCors.php",
			"line": 49,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Illuminate\\Http\\Middleware\\HandleCors",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/TrustProxies.php",
			"line": 39,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Illuminate\\Http\\Middleware\\TrustProxies",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/TrustHosts.php",
			"line": 48,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 183,
			"function": "handle",
			"class": "Illuminate\\Http\\Middleware\\TrustHosts",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
			"line": 119,
			"function": "Illuminate\\Pipeline\\{closure}",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php",
			"line": 175,
			"function": "then",
			"class": "Illuminate\\Pipeline\\Pipeline",
			"type": "->"
		},
		{
			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php",
			"line": 144,
			"function": "sendRequestThroughRouter",
			"class": "Illuminate\\Foundation\\Http\\Kernel",
			"type": "->"
		},
		{
			"file": "/var/www/html/public/index.php",
			"line": 51,
			"function": "handle",
			"class": "Illuminate\\Foundation\\Http\\Kernel",
			"type": "->"
		},
		{
			"file": "/var/www/html/server.php",
			"line": 17,
			"function": "require_once"
		}
	]
}
```

ioslife commented:
> When I call this endpoint, I get:
> 	]
> }
> ```

Huh. I'll spin it up again and see if I can get it to.

ioslife commented:
> When I call this endpoint, I get:
> 
> ```json
> {
> 	"message": "Attempt to read property \"user\" on null",
> 	"exception": "ErrorException",
> 	"file": "/var/www/html/public/API/API_GetGameLeaderboards.php",
> 	"line": 65,
> 	"trace": [
> 		{
> 			"file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php",
> 			"line": 255,
> 			"function": "handleError",
> 			"class": "Illuminate\\Foundation\\Bootstrap\\HandleExceptions",
> 			"type": "->"
> 		},
> 		{
> 			"file": "/var/www/html/public/API/API_GetGameLeaderboards.php",
> 			"line": 65,
> 			"function": "Illuminate\\Foundation\\Bootstrap\\{closure}",
> 			"class": "Illuminate\\Foundation\\Bootstrap\\HandleExceptions",
> 			"type": "->"
> 		},
>                 // ... a whole bunch of noise ...
> 	]
> }
> ```

Alright, this was caused when a leaderboard had no entries. I added a null check to fetching the topEntry data. I also added a leaderboard seeder to GamesTableSeeder.

wescopeland commented on database/seeders/GamesTableSeeder.php:
```diff
@@ -47,5 +48,13 @@ public function run(): void
+        Game::all()->each(function (Game $game) {
+            $game->leaderboards()->saveMany(Leaderboard::factory()->count(random_int(0, 10))->create([
+                'GameID' => $game->ID,
+                'Title' => "Test leaderboard",
+                'Description' => "I am a seeded leaderboard",
```
This is a bit too prescriptive -- tons of leaderboards could potentially be generated with the same titles and descriptions.

I believe the default gives us plenty of fuzz:
```php
'Title' => ucwords(fake()->words(2, true)),
'Description' => fake()->sentence(),
```

wescopeland commented on public/API/API_GetGameLeaderboards.php:
This endpoint still has a gnarly N+1 query problem. So does the UI in prod. We _really_ need to fix this in our code. I can handle the UI issue separately.

The problem in both places is caused by:
```php
$leaderboard->sortedEntries()->first();
```


https://github.com/user-attachments/assets/79cbc642-5b75-45bc-93f6-774c5a9d5958


I've spent about an hour this morning trying to come up with a good solution, and while not perfect, I believe I've found something that's better than what we're doing right now.

In _Leaderboard.php_, I've added a new scope in the "scopes" section:
```php
public function scopeWithTopEntry($query)
{
    return $query->addSelect(['top_entry_id' => function ($subQuery) {
        $subQuery->select('id')
            ->from('leaderboard_entries as le')
            ->whereColumn('le.leaderboard_id', 'LeaderboardDef.ID')
            ->orderByRaw('
                CASE 
                    WHEN LeaderboardDef.LowerIsBetter = 1 THEN le.score
                    ELSE -le.score
                END ASC
            ')
            ->orderBy('le.updated_at', 'ASC')
            ->limit(1);
    }])->with(['topEntry' => function ($query) {
        $query->with('user');
    }]);
}
```

And I've added a new relation to access this value after it's populated:
```php
public function topEntry()
{
    return $this->belongsTo(LeaderboardEntry::class, 'top_entry_id');
}
```

Now, back in the endpoint:
```php
$leaderboards = $game->leaderboards()
    ->with('game')
    ->with('developer')
    ->withTopEntry()
    ->visible()
    ->skip($offset)
    ->take($count)
    ->get();

if (!$leaderboards) {
    return response()->json([], 404);
}

$results = [];
foreach ($leaderboards as $leaderboard) {
    $topEntry = null;

    if ($leaderboard->topEntry) {
        $topEntry = [
            'User' => $leaderboard->topEntry->user->User,
            'Score' => $leaderboard->topEntry->score,
            'FormattedScore' => ValueFormat::format($leaderboard->topEntry->score, $leaderboard->Format),
        ];
    }

    $results[] = [
        'ID' => $leaderboard->ID,
        'RankAsc' => boolval($leaderboard->LowerIsBetter),
        'Title' => $leaderboard->Title,
        'Description' => $leaderboard->Description,
        'Format' => $leaderboard->Format,
        'TopEntry' => $topEntry,
    ];
}
```

For the game I'm testing with, this improves performance by 10x.

wescopeland replied:
After a bit more testing, it seems this gets close, but it doesn't handled untracked and banned users correctly.

I can modify the scope to filter those out, but I believe the ->limit(1) causes the TopEntry values to always be null if an untracked user has a top score on the leaderboard.

I'm feeling a bit stumped at the moment and need to take a break from this one.

ioslife commented on tests/Feature/Api/V1/GameLeaderboardsTest.php:
```diff
@@ -0,0 +1,266 @@
+            ->assertJson([]);
+    }
+
+    public function testGetGameLeaderboards(): void
```
That's an easy change. Will add more

ioslife replied:
I added more entries to two tests. One of the tests is the LowerIsBetter test and one is not. I don't think we need to have multiple entries in every test as having a mixture this way will give us a better look into what newer leaderboards would return.

wescopeland reviewed (changes requested):
Unfortunately, this is still not working correctly if the top score is from an untracked/unranked user.

This is tested from a recent production snapshot:
![Screenshot 2024-07-30 at 4 37 24 PM](https://github.com/user-attachments/assets/0fb48490-2e85-45d7-9731-cebdece71971)

![Screenshot 2024-07-30 at 4 38 42 PM](https://github.com/user-attachments/assets/13e7b2f1-58cf-4e9e-bd43-9f4e7fa7583f)

I would imagine for many of the top-played games on the site, leaderboard results will be filled with untracked users near the very top. Unless we find some way to fix this, it'll result in an inconsistent/unexpected results for API consumers.

There are multiple issues at play - we need to fix the N+1 problem in the UI too. Whatever solution we land on, the API layer and full-stack site layer should probably share.

One idea I have in mind to get this working is to denormalize the top leaderboard entry somewhere in the database. This reduces query complexity at the expense of calculating the top score with an async job. I'm not sure if this is the route we should go, though, it's just one idea.

wescopeland commented on app/Models/Leaderboard.php:
```diff
@@ -225,6 +233,34 @@ public function scopeVisible(Builder $query): Builder
+    public function scopeWithTopEntry(Builder $query)
+    {
+    return $query->addSelect(['top_entry_id' => function ($subQuery) {
+        $subQuery->select('le.id')
+            ->from('leaderboard_entries as le')
+            ->join('UserAccounts as u', 'u.id', '=', 'le.user_id')
+            ->whereColumn('le.leaderboard_id', 'LeaderboardDef.ID')
+            ->whereNull('u.unranked_at')
+            ->where('u.Untracked', 0)
+            ->whereNull('u.banned_at')
+            ->whereNull('le.deleted_at')
+            ->orderByRaw('
+                CASE
+                    WHEN (SELECT LowerIsBetter FROM LeaderboardDef WHERE ID = le.leaderboard_id) = 1 THEN le.score
+                    ELSE -le.score
+                END ASC
+            ')
+            ->orderBy('le.updated_at', 'ASC')
+            ->limit(1);
+    }])->with(['topEntry' => function ($query) {
+        $query->with('user');
+    }]);
+}
```
nit: There's some whitespace weirdness happening here that we should probably fix up.

ioslife replied:
See if that's better?

wescopeland commented on app/Models/Leaderboard.php:
```diff
@@ -225,6 +233,34 @@ public function scopeVisible(Builder $query): Builder
+     * @param Builder<Leaderboard> $query
+     * @return Builder<Leaderboard>
+     */
+    public function scopeWithTopEntry(Builder $query)
```
nit: Recommend also explicitly giving a static return type of `: Builder`.

wescopeland commented on app/Models/Leaderboard.php:
```diff
@@ -214,6 +214,14 @@ public function game(): BelongsTo
+    /**
+     * @return BelongsTo<LeaderboardEntry, Leaderboard>
+     */
+    public function topEntry()
```
nit: Recommend also explicitly giving a static return type of `: BelongsTo`.
